test/integration/consul-container/test/resource/resource_test.go

```diff
@@ -0,0 +1,145 @@
+// Copyright (c) HashiCorp, Inc.
+// SPDX-License-Identifier: BUSL-1.1
+
+package resource
+
+import (
+	"context"
+	"fmt"
+	"io"
+	"testing"
+
+	"github.com/stretchr/testify/require"
+
+	libcluster "github.com/hashicorp/consul/test/integration/consul-container/libs/cluster"
+	libtopology "github.com/hashicorp/consul/test/integration/consul-container/libs/topology"
+)
+
+const (
+	RESOURCE_FILE_PATH_ON_HOST      = "../../../../../command/resource/testdata/demo.hcl"
+	RESOURCE_FILE_PATH_ON_CONTAINER = "/consul/data/demo.hcl"
+)
+
+func TestClientForwardToServer(t *testing.T) {
+	type operation struct {
+		action         func(*testing.T, libcluster.Agent, string) (int, string)
+		includeToken   bool
+		expectedCode   int
+		expectedMsg    string
+		expectedErrMsg string
+	}
+	type testCase struct {
+		description string
+		operations  []operation
+		aclEnabled  bool
+	}
+
+	testCases := []testCase{
+		{
+			description: "The read request should be routed to consul server agent",
+			operations: []operation{
+				{
+					action:         applyResource,
+					includeToken:   false,
+					expectedCode:   0,
+					expectedMsg:    "demo.v2.Artist 'korn' created.",
+					expectedErrMsg: "",
+				},
+			},
+			aclEnabled: false,
+		},
+		{
+			description: "The read request should be denied if missing token when ACL is enabled",
+			operations: []operation{
+				{
+					action:         applyResource,
+					includeToken:   false,
+					expectedCode:   1,
+					expectedMsg:    "",
+					expectedErrMsg: "failed getting authorizer: ACL not found",
+				},
+				{
+					action:         applyResource,
+					includeToken:   true,
+					expectedCode:   0,
+					expectedMsg:    "demo.v2.Artist 'korn' created.",
+					expectedErrMsg: "",
+				},
+			},
+			aclEnabled: true,
+		},
+	}
+
+	for _, tc := range testCases {
+		tc := tc
+		t.Run(tc.description, func(t *testing.T) {
+			t.Parallel()
+
+			var clientAgent libcluster.Agent
+			cluster, clientAgent := setupClusterAndClient(t, tc.aclEnabled)
+			defer terminate(t, cluster)
+
+			// perform actions and validate returned messages
+			for _, op := range tc.operations {
+				token := ""
+				if op.includeToken {
+					token = cluster.TokenBootstrap
+				}
+				code, res := op.action(t, clientAgent, token)
+				require.Equal(t, op.expectedCode, code)
+				if code == 0 {
+					require.Contains(t, res, op.expectedMsg)
+				} else {
+					require.Contains(t, res, op.expectedErrMsg)
+				}
```

Comment:
Prefer a single collapsed `expectedMsg` since `code` asserts the error condition is correct.

Reply:
yup, looks better, updated